src/middlewared/middlewared/service/sharing_service.py

```diff
@@ -24,17 +24,19 @@ async def get_path_field(self, data):
 
     @private
     async def sharing_task_extend_context(self, rows, extra):
-        datasets = sum([
-            await self.middleware.call(f'{self._config.namespace}.sharing_task_datasets', row)
-            for row in rows
-        ], [])
-
+        if extra.get('use_cached_locked_datasets', True):
+            locked_ds_endpoint = 'pool.dataset.locked_datasets_cached'
+        else:
+            locked_ds_endpoint = 'zfs.dataset.locked_datasets'
         return {
-            'locked_datasets': await self.middleware.call('zfs.dataset.locked_datasets', datasets) if datasets else [],
+            'locked_datasets': await self.middleware.call(
+                locked_ds_endpoint
+            ) if extra.get('retrieve_locked_info', True) else [],
```

Comment:
I think instead of using `extra.retrieve_locked_info` it would be easier and less error-prone to omit locked info lookup in cases where `select` is specified and doesn't include `locked`. We'd have to update existing callers, but this would hopefully make future code using fast-path easier to write.